### PR TITLE
Improve the grammar of numeric literals

### PR DESCRIPTION
**Pull request by esdoppio**
Please merge #155 before this :wink:

With atom-material-syntax, now:

578a9d8c-4f7b-45d4-b89c-b6e46e149938

New:

fe194fa4-dfcb-4181-a8a4-e28e38d5932a

In Ruby, for `-(1)` and `- 1` (note the space in between), `-` is parsed as an operator (a method of Fixnum under the hood), but for `-1`, `-` is parsed as part of the literal~~, so `-` should be tokenized accordingly in the grammar~~.

### CODE REVIEW

**Comment by 50Wliu** [comment]
Rebase please!

File: grammars/ruby.cson [hunk 1]
```diff
@@ -351,7 +351,8 @@
   }
   {
     'match': '''(?x)
-      \\b(
+      (\\b|(?<!\\d[ \\t]|\\d)-)
```
Comment (50Wliu) [inline grammars/ruby.cson]: In Ruby is the following valid?
```ruby
3
-4
```
Is that equivalent to `3 - 4`?

Previously we've had problems trying to include the negative sign as part of the number.

Comment (esdoppio) [inline grammars/ruby.cson]: Not equivalent this way. You have to tell the parser this line ain't over by placing a backslash at the end:

```Ruby
3 \
-4
```

Yet the rule still breaks. `3  -4` also breaks it. But seriously, for a person who writes code this way, I think they wouldn't care. 😂  Could you show me where did the problems happen?

On second thought, `(\\b|(?<!\\w|\\w[ \\t])-)` is better, or just `(\\b|(?<!\\w)-)`, if we are still gonna keep this commit.

Comment (50Wliu) [inline grammars/ruby.cson]: Both of those have the same problems you just demonstrated with the double spaces.  It also gets more complicated with parentheses: `(3) - 4`.  So I think reverting 0a886fc is the correct choice here, as `-` is still technically an operator in these cases.

File: grammars/ruby.cson [hunk 2]
```diff
@@ -355,13 +355,13 @@
...
-        0[xX]\\h(?>_?\\h)*
-        |
-        0[oO]?[0-7](?>_?[0-7])*
-        |
-        0[bB][01](?>_?[01])*
-        |
-        0[dD]\\d(?>_?\\d)*                          # A base indicator can only be used with an integer
+        0
+        (?:
+          [xX]\\h(?>_?\\h)*|
```
Comment (esdoppio) [inline grammars/ruby.cson]: In this case, each line in the group is an alternative, so I chose to put the `|` at the end. 😉  I'll be consistent, promise.

File: grammars/ruby.cson [hunk 3]
```diff
@@ -360,6 +360,8 @@
         0[oO]?[0-7](?>_?[0-7])*
         |
         0[bB][01](?>_?[01])*
+        |
+        0[dD]\\d(?>_?\\d)*                          # A base indicator can only be used with an integer
```
Comment (esdoppio) [inline grammars/ruby.cson]: I could instead add a `(?:0[dD])?` to the beginning of L354, but in this way right now, when people write something like `0d1.2`, they could get a hint that it's not a valid number.

File: grammars/ruby.cson [hunk 4]
```diff
@@ -351,7 +351,7 @@
     'match': '''(?x)
       \\b
       (
-        \\d(?>_?\\d)*                               # 100_000
+        [1-9](?>_?\\d)*                             # 100_000
```
Comment (esdoppio) [inline grammars/ruby.cson]: Similarly, when people write something like `08`, they'll get a hint.

File: spec/ruby-spec.coffee [hunk 5]
```diff
@@ -32,6 +32,13 @@ describe "Ruby grammar", ->
     expect(tokens[1]).toEqual value: '&.', scopes: ['source.ruby', 'punctuation.separator.method.ruby']
     expect(tokens[2]).toEqual value: 'call', scopes: ['source.ruby']
 
+  it "tokenizes octal numbers", ->
```
Comment (50Wliu) [inline spec/ruby-spec.coffee]: Can tests for all the other number types also be added?

**Comment by 50Wliu** [comment]
One more thing.  Can each number type get its own subscope like `constant.numeric.hex.ruby` or `constant.numeric.octal.ruby`?

Once again referring to https://github.com/atom/language-javascript/blob/74ff62f9a93de2c1976432feccce2d8ee23aa740/grammars/javascript.cson#L1076.

**Comment by esdoppio** [comment]
Yes they could, but I suggest we follow the "You Ain't Gonna Need It" practice, and only add those when we actually need them. 😉

**Comment by 50Wliu** [comment]
\<whine\> "BUT I WANT THEM" \</whine\>

:laughing:

**Comment by esdoppio** [comment]
😆  Go get 'em!

**Pull request merged**